models/cm_tree.cpp

```diff
@@ -0,0 +1,487 @@
+fault
+*/
+nest::Compartment*
+nest::CompTree::get_compartment_opt( const long compartment_idx ) const
+{
+  return compartments_[ compartment_idx ];
+}
+
+/*
+Initialize all tree structure pointers
+*/
+void
+nest::CompTree::init_pointers()
+{
+  set_parents();
+  set_compartments();
+  set_leafs();
+}
+/*
+For each compartments, sets its pointer towards its parent compartment
+*/
+void
+nest::CompTree::set_parents()
+{
+  for ( auto compartment_idx_it = compartment_indices_.begin(); compartment_idx_it != compartment_indices_.end();
+        ++compartment_idx_it )
+  {
+    Compartment* comp_ptr = get_compartment( *compartment_idx_it );
+    // will be nullptr if root
+    Compartment* parent_ptr = get_compartment( comp_ptr->p_index, &root_, 0 );
+    comp_ptr->parent = parent_ptr;
+  }
+}
+/*
+Creates a vector of compartment pointers, organized in the order in which they were
+added by `add_compartment()`
+*/
+void
+nest::CompTree::set_compartments()
+{
+  compartments_.clear();
+
+  for ( auto compartment_idx_it = compartment_indices_.begin(); compartment_idx_it != compartment_indices_.end();
+        ++compartment_idx_it )
+  {
+    compartments_.push_back( get_compartment( *compartment_idx_it ) );
+  }
+}
+/*
+Creates a vector of compartment pointers of compartments that are also leafs of the tree.
+*/
+void
+nest::CompTree::set_leafs()
+{
+  leafs_.clear();
+  for ( auto compartment_it = compartments_.begin(); compartment_it != compartments_.end(); ++compartment_it )
+  {
+    if ( int( ( *compartment_it )->children.size() ) == 0 )
+    {
+      leafs_.push_back( *compartment_it );
+    }
+  }
+}
+
+/*
+Initializes pointers for the spike buffers for all synapse receptors
+*/
+void
+nest::CompTree::set_syn_buffers( std::vector< RingBuffer >& syn_buffers )
+{
+  for ( auto compartment_it = compartments_.begin(); compartment_it != compartments_.end(); ++compartment_it )
+  {
+    ( *compartment_it )->compartment_currents.set_syn_buffers( syn_buffers );
+  }
+}
+
+/*
+Returns a map of variable names and pointers to the recordables
+*/
+std::map< Name, double* >
+nest::CompTree::get_recordables()
+{
+  std::map< Name, double* > recordables;
+
+  /*
+  add recordables for all compartments, suffixed by compartment_idx,
+  to "recordables"
+  */
+  for ( auto compartment_it = compartments_.begin(); compartment_it != compartments_.end(); ++compartment_it )
+  {
+    std::map< Name, double* > recordables_comp = ( *compartment_it )->get_recordables();
+    recordables.insert( recordables_comp.begin(), recordables_comp.end() );
+  }
+  return recordables;
+}
+
+/*
+Initialize state variables
+*/
+void
+nest::CompTree::calibrate()
+{
+  if ( root_.comp_index < 0 )
+  {
+    std::string msg = "does not exist in tree, meaning that no compartments have been added";
+    throw UnknownCompartment( 0, msg );
+  }
+
+  // initialize the compartments
+  for ( auto compartment_it = compartments_.begin(); compartment_it != compartments_.end(); ++compartment_it )
```

Comment:
Maybe convert to a full [range-based C++ loop](https://en.cppreference.com/w/cpp/language/range-for)? This comment also applies in other places.